### PR TITLE
Prl functional test

### PR DESCRIPTION
**Pull request by sahitya-desireddy**
Functional Test Cases Enhancement

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/AddCafcassOfficerControllerFunctionalTest.java [hunk 1]
```diff
@@ -40,24 +32,40 @@ public class AddCafcassOfficerControllerFunctionalTest {
...
+        StringUtils.defaultIfBlank(
+            System.getenv("TEST_URL"),
+            "http://localhost:4044"
+        );
+
+    private final RequestSpecification request = RestAssured.given().relaxedHTTPSValidation().baseUri(targetInstance);
 
     @Test
     public void givenRequestBody_whenAdd_cafcass_officer_about_to_submit_then200Response() throws Exception {
         String requestBody = ResourceLoader.loadJson(VALID_REQUEST_BODY);
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/AddCafcassOfficerControllerFunctionalTest.java:48]: Are we not creating a fresh case?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/C100IssueCaseControllerFunctionalTest.java [hunk 2]
```diff
@@ -1,76 +1,84 @@
...
-        String requestBody = ResourceLoader.loadJson(VALID_REQUEST_BODY);
-        mockMvc.perform(post("/issue-and-send-to-local-court")
-                            .contentType(MediaType.APPLICATION_JSON)
-                            .header("Authorization", idamTokenGenerator.generateIdamTokenForSolicitor())
-                            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
-                            .content(requestBody)
-                            .accept(MediaType.APPLICATION_JSON))
-            .andExpect(status().isOk())
-            .andReturn();
+        String requestBody = ResourceLoader.loadJson(VALID_CAFCASS_REQUEST_JSON);
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/C100IssueCaseControllerFunctionalTest.java:44]: fresh case should be created here?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/C100IssueCaseControllerFunctionalTest.java [hunk 3]
```diff
@@ -1,76 +1,84 @@
...
+        request
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForCitizen())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType("application/json")
+            .post("/issue-and-send-to-local-court")
+            .then()
+            .assertThat().statusCode(200)
+            .body("data.caseStatus.state", equalTo("Gatekeeping"),
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/C100IssueCaseControllerFunctionalTest.java:55]: Should be issued state?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/C100IssueCaseControllerFunctionalTest.java [hunk 4]
```diff
@@ -1,76 +1,84 @@
...
-                            .accept(MediaType.APPLICATION_JSON))
-            .andExpect(status().isOk())
-            .andReturn();
+        request
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForCitizen())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType("application/json")
+            .post("/issue-and-send-to-local-court-notification")
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/C100IssueCaseControllerFunctionalTest.java:76]: Can we verify, how many times a service was called?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java [hunk 5]
```diff
@@ -145,284 +49,258 @@ public class CallbackControllerFT {
...
+            .header("Content-Type", APPLICATION_JSON_VALUE)
+            .header("Accepts", APPLICATION_JSON_VALUE)
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType(APPLICATION_JSON_VALUE)
+            .post("/update-application")
+            .then()
+            .assertThat().statusCode(200);
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java:144]: Can we assert the data here?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java [hunk 6]
```diff
@@ -145,284 +49,258 @@ public class CallbackControllerFT {
...
+            .header("Accepts", APPLICATION_JSON_VALUE)
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType(APPLICATION_JSON_VALUE)
+            .post("/case-withdrawn-about-to-submit")
+            .then()
+            .assertThat().statusCode(200)
+            .body("data.welshLanguageRequirementsTable", notNullValue(),
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java:162]: is not null sufficient?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java [hunk 7]
```diff
@@ -145,284 +49,258 @@ public class CallbackControllerFT {
...
+            .header("Accepts", APPLICATION_JSON_VALUE)
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType(APPLICATION_JSON_VALUE)
+            .post("/send-to-gatekeeper")
+            .then()
+            .assertThat().statusCode(200)
+            .body("data.gatekeepingDetails", notNullValue());
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java:186]: is not null sufficient?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java [hunk 8]
```diff
@@ -145,284 +49,258 @@ public class CallbackControllerFT {
...
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType(APPLICATION_JSON_VALUE)
+            .post("/update-party-details")
+            .then()
+            .assertThat().statusCode(200)
+            .body("data.caseTypeOfApplication", equalTo("C100"),
+                  "data.caApplicant1Policy.OrgPolicyCaseAssignedRole",  equalTo("[C100APPLICANTSOLICITOR1]"),
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java:205]: What are we asserting here? and the name of the method has any relation with this?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java [hunk 9]
```diff
@@ -145,284 +49,258 @@ public class CallbackControllerFT {
...
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType(APPLICATION_JSON_VALUE)
+            .post("/pre-populate-court-details")
+            .then()
+            .assertThat().statusCode(200)
+            .body("data.caseTypeOfApplication", equalTo("C100"));
+
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java:263]: What are we achieving in this and previous test case?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java [hunk 10]
```diff
@@ -145,284 +49,258 @@ public class CallbackControllerFT {
...
+            .header("Accepts", APPLICATION_JSON_VALUE)
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType(APPLICATION_JSON_VALUE)
+            .post("/send-to-gatekeeper")
+            .then()
+            .assertThat().statusCode(200)
+            .body("data.caseTypeOfApplication", equalTo("C100"),
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java:282]: Should be asserting the gatekeeper details

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java [hunk 11]
```diff
@@ -439,9 +317,7 @@ public void testAttachScanDocsWaChange() throws Exception {
             .post("/attach-scan-docs/about-to-submit")
             .then()
             .body("data.manageDocumentsRestrictedFlag", equalTo(TRUE),
-                  "data.manageDocumentsTriggeredBy", equalTo(STAFF))
-            .extract()
-            .as(AboutToStartOrSubmitCallbackResponse.class);
+                  "data.manageDocumentsTriggeredBy", equalTo(STAFF));
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFT.java:320]: It should be bulk scan and not cout staff as scanning will be done by bulk scan service

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFunctionalTest.java [hunk 12]
```diff
@@ -183,14 +170,18 @@ public void givenRequestWithC100ApplicantDetails_whenEndPointCalled_ResponseCont
...
+        request
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSolicitor())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType("application/json")
+            .post("/fl401-add-case-number")
+            .then()
+            .assertThat().statusCode(200)
+            .body("data.issueDate", notNullValue(),
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFunctionalTest.java:182]: notNullValue() is that enough?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/ConfidentialityCheckControllerFT.java [hunk 13]
```diff
@@ -95,34 +76,37 @@ public void createCcdTestCase() throws Exception {
...
+        request
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType("application/json")
+            .post("/confidentiality-check/about-to-start")
+            .then()
+            .assertThat().statusCode(200)
+            .body("data.unServedApplicantPack.packDocument", nullValue(),
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/ConfidentialityCheckControllerFT.java:88]: what is it doing?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/ConfidentialityCheckControllerFT.java [hunk 14]
```diff
@@ -267,17 +251,15 @@ public void givenRequestWithCaseData_ResponseContainsYes() throws Exception {
...
-                                            .contentType(MediaType.APPLICATION_JSON)
-                                            .content(requestBody)
-                                            .accept(MediaType.APPLICATION_JSON))
-            .andExpect(status().isOk())
-            .andReturn();
-
-        String json = res.getResponse().getContentAsString();
-        assertTrue(json.contains("confirmation_header"));
+        request
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/ConfidentialityCheckControllerFT.java:255]: Should be manager user.

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/DraftOrdersControllerFunctionalTest.java [hunk 15]
```diff
@@ -79,112 +42,108 @@ public class DraftOrdersControllerFunctionalTest {
...
+        request
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType("application/json")
+            .post("/selected-order")
+            .then()
+            .assertThat().statusCode(200)
+            .body("errors[0]", equalTo("This order is not available to be drafted"));
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/DraftOrdersControllerFunctionalTest.java:73]: WE are checking for negative scenario and can we add test case for positive scenario? When correct order is selected should be asserted

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/DraftOrdersControllerFunctionalTest.java [hunk 16]
```diff
@@ -79,112 +42,108 @@ public class DraftOrdersControllerFunctionalTest {
...
+            .when()
+            .contentType("application/json")
+            .post("/populate-draft-order-fields")
+            .then()
+            .assertThat().statusCode(200)
+            .body("data.caseTypeOfApplication", equalTo("FL401"),
+                  "data.ordersHearingDetails[0].value.hearingTypes.value", notNullValue(),
+                  "data.draftOrderCollection[0].id", notNullValue(),
+                  "data.draftOrdersDynamicList", notNullValue()
+                  );
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/DraftOrdersControllerFunctionalTest.java:94]: Non null is not sufficient

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/DraftOrdersControllerFunctionalTest.java [hunk 17]
```diff
@@ -79,112 +42,108 @@ public class DraftOrdersControllerFunctionalTest {
...
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType("application/json")
+            .post("/populate-standard-direction-order-fields")
+            .then()
+            .assertThat().statusCode(200)
+            .body("data.caseTypeOfApplication", equalTo("FL401"),
+                  "data.draftOrderCollection[0].id", notNullValue());
+
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/DraftOrdersControllerFunctionalTest.java:112]: Weher is the assertion for SDO fields?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/DraftOrdersControllerFunctionalTest.java [hunk 18]
```diff
@@ -79,112 +42,108 @@ public class DraftOrdersControllerFunctionalTest {
...
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType("application/json")
+            .post("/about-to-submit")
+            .then()
+            .assertThat().statusCode(200)
+            .body("data.caseTypeOfApplication", equalTo("FL401"),
+                  "data.caseStatus.state", equalTo("Draft"));
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/DraftOrdersControllerFunctionalTest.java:128]: We should be checking the draft order state and not case state

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/DraftOrdersControllerFunctionalTest.java [hunk 19]
```diff
@@ -79,112 +42,108 @@ public class DraftOrdersControllerFunctionalTest {
...
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType("application/json")
+            .post("/generate-doc")
+            .then()
+            .assertThat().statusCode(200)
+            .body("errors", nullValue(),
+                  "data.caseTypeOfApplication", equalTo("FL401"));
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/DraftOrdersControllerFunctionalTest.java:145]: what is it doing?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/c100respondentsolicitor/C100RespondentSolicitorControllerFunctionalTest.java [hunk 20]
```diff
@@ -99,95 +76,125 @@ public void createCcdTestCase() throws Exception {
...
+            .body(requestBody)
+            .when()
+            .contentType("application/json")
+            .post("/respondent-solicitor/about-to-submit")
+            .then()
+            .extract()
+            .as(AboutToStartOrSubmitCallbackResponse.class);
+        Assert.assertNotNull(responseData);
+        Assert.assertNotNull(responseData.getData().get("solicitorName"));
+        Assert.assertEquals("AAT Solicitor",responseData.getData().get("solicitorName"));
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/c100respondentsolicitor/C100RespondentSolicitorControllerFunctionalTest.java:114]: in about to start and about to submit we are asserting the same.

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/c100respondentsolicitor/C100RespondentSolicitorControllerFunctionalTest.java [hunk 21]
```diff
@@ -99,95 +76,125 @@ public void createCcdTestCase() throws Exception {
...
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType("application/json")
+            .post("/respondent-solicitor/about-to-start-response-validation")
+            .then()
+            .extract()
+            .as(AboutToStartOrSubmitCallbackResponse.class);
+        Assert.assertNotNull(responseData);
+        Assert.assertNotNull(responseData.getData().get("solicitorName"));
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/c100respondentsolicitor/C100RespondentSolicitorControllerFunctionalTest.java:132]: same as before

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/c100respondentsolicitor/C100RespondentSolicitorControllerFunctionalTest.java [hunk 22]
```diff
@@ -99,95 +76,125 @@ public void createCcdTestCase() throws Exception {
...
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(requestBody)
+            .when()
+            .contentType("application/json")
+            .post("/respondent-solicitor/submit-c7-response")
+            .then()
+            .extract()
+            .as(AboutToStartOrSubmitCallbackResponse.class);
+        Assert.assertNotNull(responseData);
+        Assert.assertNotNull(responseData.getData().get("solicitorName"));
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/c100respondentsolicitor/C100RespondentSolicitorControllerFunctionalTest.java:151]: same as before

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/citizen/CitizenCallbackControllerFunctionalTest.java [hunk 23]
```diff
@@ -1,73 +1,79 @@
...
+            .header("Content-Type", APPLICATION_JSON_VALUE)
+            .header("Accepts", APPLICATION_JSON_VALUE)
+            .header("Authorization", idamTokenGenerator.generateIdamTokenForSystem())
+            .header("ServiceAuthorization", serviceAuthenticationGenerator.generateTokenForCcd())
+            .body(CallbackRequest.builder().caseDetails(caseDetails).build())
+            .when()
+            .contentType(APPLICATION_JSON_VALUE)
+            .post("/update-citizen-application")
+            .then()
+            .assertThat().statusCode(200);
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/citizen/CitizenCallbackControllerFunctionalTest.java:75]: Assert citizen fields

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/noticeofchnage/NoticeOfChangeControllerFunctionalTest.java [hunk 24]
```diff
@@ -38,48 +41,65 @@ public class NoticeOfChangeControllerFunctionalTest {
...
+            .when()
+            .contentType(APPLICATION_JSON_VALUE)
             .post("/noc/aboutToSubmitNoCRequest")
             .then()
-            .statusCode(200);
+            .assertThat().statusCode(200)
+            .extract()
+            .as(AboutToStartOrSubmitCallbackResponse.class);
+
+        Assert.assertNotNull(response1.getData());
```
Comment (shashisk97) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/noticeofchnage/NoticeOfChangeControllerFunctionalTest.java:84]: Should we assert org policy?

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/AddCafcassOfficerControllerFunctionalTest.java (hunk 1 above)
Comment (sahitya-desireddy) [reply src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/AddCafcassOfficerControllerFunctionalTest.java:48]: Tried to create a fresh case but isAuthorised function in the method throwing error due to use of authorisation tokens twice , tried a day to find the root cause of the error , hence using the json input for all the methods using isAuthorise function .... fyi @shashisk97 @yogendraU

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/C100IssueCaseControllerFunctionalTest.java (hunk 2 above)
Comment (sahitya-desireddy) [reply src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/C100IssueCaseControllerFunctionalTest.java:44]: Tried to create a fresh case but isAuthorised function in the method throwing error due to use of authorisation tokens twice , tried a day to find the root cause of the error , hence using the json input for all the methods using isAuthorise function .... fyi @shashisk97 @yogendraU

File: src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFunctionalTest.java (hunk 12 above)
Comment (yogendraU) [inline src/functionalTest/java/uk/gov/hmcts/reform/prl/controllers/CallbackControllerFunctionalTest.java:182]: also this event moves case state

**Review by AlokDatta — APPROVED** [review]

**Pull request merged**